File: Backend/Tictactoe/Models/Match.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using Tictactoe.Types.Enums;

namespace Tictactoe.Models;

public class Match
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string? Id { get; set; }

    [BsonElement("name")]
    [BsonRepresentation(BsonType.String)]
    public string Name { get; set; } = null!;

    [BsonElement("hashedPassword")] 
    [BsonRepresentation(BsonType.String)]
    public string HashedPassword { get; set; } = null!;

    [BsonElement("salt")]
    [BsonRepresentation(BsonType.String)]
    public string Salt { get; set; } = null!;

    [BsonElement("creatorId")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string CreatorId { get; set; } = null!;

    [BsonElement("memberId")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string? MemberId { get; set; } = null!;

    [BsonElement("creatorStatus")]
    public PlayerStatus? CreatorStatus { get; set; } = PlayerStatus.Left;

    [BsonElement("memberStatus")]
    public PlayerStatus? MemberStatus { get; set; } = PlayerStatus.Left;

    [BsonElement("creatorMoves")]
    public ushort CreatorMoves { get; set; } = 0;

    [BsonElement("memberMoves")]
    public ushort MemberMoves { get; set; } = 0;

    [BsonElement("nextTurn")]
    public GameTurn NextTurn { get; set; } = GameTurn.Creator;

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("updatedAt")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("gameOutcome")]
    public GameOutcome GameOutcome { get; set; } = GameOutcome.Going;

    public bool HasFinished
    {
        get
        {
            return GameOutcome != default && GameOutcome != GameOutcome.Going;
        }
    }

    public bool HashMember(string userId)
    {
        return string.Equals(MemberId, userId, StringComparison.Ordinal) || string.Equals(CreatorId, userId, StringComparison.Ordinal);
    }
}

File: Frontend/src/types/match.ts
export const ePlayerStatus = Object.freeze({
  Joined: 0,
  Left: 1,
});

type PlayerStatus = (typeof ePlayerStatus)[keyof typeof ePlayerStatus];

export const eGameTurn = Object.freeze({
  Creator: 0,
  Member: 1,
});

export type GameTurn = (typeof eGameTurn)[keyof typeof eGameTurn];

export const eGameOutcome = Object.freeze({
  Going: -1,
  Draw: 0,
  CreatorWin: 1,
  PlayerWin: 2,
});

export type GameOutcome = (typeof eGameOutcome)[keyof typeof eGameOutcome];

export const eRoomActivity = Object.freeze({
  MakeMove: 0,
  LeaveRoom: 1,
  JoinRoom: 2,
});

type RoomActivity = (typeof eRoomActivity)[keyof typeof eRoomActivity];

export interface Match {
  id: string;
  name: string;
  creatorId: string;
  memberId: string;
  creatorConnectionId: string;
  memberConnectionId: string;
  creatorStatus: PlayerStatus;
  memberStatus: PlayerStatus;
  creatorMoves: number;
  memberMoves: number;
  nextTurn: GameTurn;
  createdAt: string;
  updatedAt: string;
  gameOutcome: GameOutcome;
  hasFinished: boolean;
  isBlocked: boolean;
}

export interface MatchResults {
  matches: Match[];
  count: number;
}

export interface RoomActivityUpdateResponse {
    match?: Match | null,
    error?: {
        errorMessage: string;
        errorCode: string;
        errorDetailMessage: string;
    }
}

export interface RoomActivityUpdateRequest {
    roomId: string;
    roomActivity: RoomActivity;
    move?: number;
}

export interface CreateMatchRequest {
  name: string;
}

export interface WSInvokeOutput<T> {
  isCanceled: boolean;
  isCompleted: boolean;
  isCompletedSuccessfully: boolean;
  isFaulted: boolean;
  result: T | null;
}
